design/src/rfcs/rfc0016_body_callback_apis_v2.md

```diff
@@ -231,19 +254,24 @@ impl ChecksumCallback {
         }
     }
 
-    // TODO I think I still need to account for the size of the trailer names.
-    //      Should I be doing that here?
     pub fn size_hint(&self) -> SizeHint {
-        let checksum_size_in_bytes = match &self.0 {
-            Inner::Crc32(_) => Crc32Callback::size(),
-            Inner::Crc32c(_) => Crc32cCallback::size(),
-            Inner::Sha1(_) => Sha1Callback::size(),
-            Inner::Sha256(_) => Sha256Callback::size(),
+        let (trailer_name_size_in_bytes, checksum_size_in_bytes) = match &self.0 {
+            // We want to get the size of the actual `HeaderName` except those don't have a `.len()`
+            // method so we'd have to convert back to the original string. That's why we're getting
+            // `.len()` from the original strings here.
+            Inner::Crc32(_) => (CRC_32_HEADER_NAME.len(), Crc32Callback::size()),
```

Comment:
It seems like these should use the `header_name` function on the callback. Perhaps this could be simplified more if a trait is added for callback, and `Inner` has functions `callback` and `callback_mut` that return `&dyn Callback` and `&mut dyn Callback` respectively. With the trait, you could keep the switching logic limited to the constructor and to these function impls in `Inner`, rather than in every place the callback is referenced.